src/deploy_pex.py

```diff
@@ -0,0 +1,201 @@
+# ubuntu-20.04 can always build pexes that work on our target platform
+# ubuntu-22.04 can only build pexes if there are no sdists (source only packages)
+
+# On ubuntu-20.04: forward args to builder.pex
+
+# On ubuntu-22.04:
+#   Try to build without sdists
+#     If success: all good (no sdists were involved)
+#     If failure: run builder.pex within the dagster-manylinux-builder docker image
+
+import os
+from pathlib import Path
+import subprocess
+import sys
+
+BUILDER_PEX_PATH = Path(__file__).parent.parent / "generated/gha/builder.pex"
+DOCKER_PATH = "/usr/bin/docker"
+DOCKER_IMAGE = "ghcr.io/dagster-io/dagster-manylinux-builder:v0.1"
+
+
+def main():
+    args = sys.argv[1:]
+    ubuntu_version = get_runner_ubuntu_version()
+    print("Running on Ubuntu", ubuntu_version)
+    if ubuntu_version == "20.04":
+        returncode, output = deploy_pex_from_current_environment(args, build_sdists=True)
+    else:
+        returncode, output = deploy_pex_from_current_environment(args, build_sdists=False)
+        if returncode:
+            dep_failures = dependency_failure_lines(output)
+            if dep_failures:
+                try:
+                    print(
+                        "::group::Preparing a Docker build environment to build the PEX files",
+                        flush=True,
+                    )
+                    print(
+                        "Going to build dependencies within a Docker build environment "
+                        "to resolve missing binary packages for the following:"
+                    )
+                    for line in dep_failures:
+                        print(f"- {line}")
+
+                    returncode, output = deploy_pex_from_docker(args)
+                finally:
+                    print("::endgroup::", flush=True)
+
+    if returncode:
+        print(
+            "::error Title=Deploy failed::Failed to deploy Python Executable. "
+            "Try removing `ENABLE_FAST_DEPLOYS: 'true'` from your .github/workflows/*yml."
+        )
+        # TODO: fallback to docker deploy here
+        sys.exit(1)
+
+
+def get_runner_ubuntu_version():
+    release_info = open("/etc/lsb-release", encoding="utf-8").read()
+    # Example:
+    # DISTRIB_ID=Ubuntu
+    # DISTRIB_RELEASE=22.04
+    # DISTRIB_CODENAME=jammy
+    # DISTRIB_DESCRIPTION="Ubuntu 22.04.1 LTS"
+    for line in release_info.splitlines(keepends=False):
+        if line.startswith("DISTRIB_RELEASE="):
+            return line.split("=", 1)[1]
+    return "22.04"  # fallback to safer behavior
+
+
+def run(args):
+    # Prints streaming output and also captures and returns it
+    print("Running", args)
+    popen = subprocess.Popen(
+        args, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, encoding="utf-8"
+    )
+    output = []
+    for line in iter(popen.stdout.readline, ""):
+        print(line, end="")
+        output.append(line)
+    popen.stdout.close()
+    returncode = popen.wait()
```

Comment:
my understanding is that this can deadlock if you're not careful? Should we use communicate() instead as they recommend here? https://docs.python.org/3/library/subprocess.html#subprocess.Popen.stdout

Reply:
Since this only reads and doesn't write to the process (no `proc.stdin`) a deadlock shouldn't happen. IIUC deadlocks happen when we're waiting on the subprocess to empty it's stdin buffer while it is waiting for us to empty it's stdout/stderr buffer. Since we are continuously emptying its stdout/stderr buffer, this should be good.

The problem with `communicate()` is it doesn't do streaming and the output goes silent if you're tailing the logs.